### PR TITLE
[TASK -1] Setup Project with CI/CD

### PR DESCRIPTION
**Pull request by rafayet-monon**
## What happened

 **[Feature Card](https://github.com/rafayet-monon/surveyor/projects/1#card-43667296)**

- Setup project
- Configure CI/CD

## Insight

- [x] Create a new project with create-react-app
- [x] Setup ES linting
- [x] Setup SASS and SASS linting
- [x] Setup prettier
- [x] Setup folder structure
- [x] Setup Heroku App
- [x] Setup Github actions CI for pull request
- [x] Setup Github actions CD for deployment
- [x] Edit Readme with proper instructions
- [x] Implement Danger System

## Proof Of Work

- **Project Structure -** 

Screenshot 2020-08-15 at 3 55 27 PM

-  **Continuous Integration -** 

This Pull Request will pass with ES lint, SASS lint and Test Suites .

Screenshot 2020-08-15 at 4 10 09 PM

-  **Continuous Deployment -** 

Screenshot 2020-08-15 at 3 42 06 PM

-  **Danger System -** 

There should be a [**comment by Danger**](https://github.com/rafayet-monon/surveyor/pull/1#issuecomment-674546203) when added a faulty commit and CI should pass if everything is okay.

### CODE REVIEW

File: src/App.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.Test {
```
Comment (rafayetn) [inline src/App.scss]: - :warning: Selector should be written in lowercase with hyphens (selector-class-pattern)

**Comment by rafayetn** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  
<b>🃏 FAIL</b> in <a href="https://github.com/rafayet-monon/surveyor/blob/chore/setup-project/src/App.test.js">App.test.js</a>


<li>
<a href='https://github.com/rafayet-monon/surveyor/blob/chore/setup-project/src/App.test.js#L7'>renders app name SURVEYOR</a>
<br/>
<body> <div> <div class="App" > <header class="App-header" > <p> SURVEYOR </p> </header> </div> </div></body>

<details>
<summary>Full message</summary>
<pre><code>
Error: Unable to find an element with the text: /SURVEYOR_TEST/i. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.

<body>
  <div>
    <div
      class="App"
    >
      <header
        class="App-header"
      >
        <p>
          SURVEYOR
        </p>
      </header>
    </div>
  </div>
</body>
    at Object.getElementError (/home/runner/work/surveyor/surveyor/node_modules/@testing-library/dom/dist/config.js:34:12)
    at /home/runner/work/surveyor/surveyor/node_modules/@testing-library/dom/dist/query-helpers.js:71:38
    at getByText (/home/runner/work/surveyor/surveyor/node_modules/@testing-library/dom/dist/query-helpers.js:54:17)
    at Object.<anonymous> (/home/runner/work/surveyor/surveyor/src/App.test.js:7:19)
    at Object.asyncJestTest (/home/runner/work/surveyor/surveyor/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)
    at /home/runner/work/surveyor/surveyor/node_modules/jest-jasmine2/build/queueRunner.js:43:12
    at new Promise (<anonymous>)
    at mapper (/home/runner/work/surveyor/surveyor/node_modules/jest-jasmine2/build/queueRunner.js:26:19)
    at /home/runner/work/surveyor/surveyor/node_modules/jest-jasmine2/build/queueRunner.js:73:41
</code></pre>
</details>
</li>
<br/>


  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Eslint Failure: Path: /home/runner/work/surveyor/surveyor/src/App.js - Line: 3 - message: 'unused_eslint' is assigned a value but never used.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  **src/App.scss#L2** - Unexpected named color "white" (color-named)
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 585f50b7416fad0c1b19245b8376bf80ee0c7740
</p>

File: .github/workflows/continuous-integration.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v1
+
+      - name: Install node packages
+        run: npm install
+
+      - name: Run Eslint
+        if: always()
```
Comment (namtx) [inline .github/workflows/continuous-integration.yml]: just curious, could the `if: always()` be omitted?

File: dangerfile.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+const stylelint = require('danger-plugin-stylelint')
+
+// ESLINT
+const esLintFile = "eslint-output.json";
+const esLinterOutput = fs.readFileSync(esLintFile).toString();
+const esLintJson = JSON.parse(esLinterOutput);
+
+if (Object.keys(esLintJson).length !== 0) {
+  for (let i = 0; i < esLintJson.length; i++) {
+    let lint_obj = esLintJson[i];
```
Comment (namtx) [inline dangerfile.js]: prefer `const` over `let` I think

File: src/App.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import './App.scss';
+
+function App() {
+  return (
+    <div className="App">
+      <header className="App-header">
```
Comment (namtx) [inline src/App.js]: about CSS, I think we can follow our guideline on the [compass](https://compass.nimblehq.co/development/code-conventions/css/)

File: src/App.test.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+import App from './App';
+
+test('renders app name SURVEYOR', () => {
```
Comment (namtx) [inline src/App.test.js]: Should we follow the template:
```js
describe('whatever', () => {
  context('whatever', () => {
    it('renders something', () => {
       // expect
    })
  })
})
```

**Review by namtx — CHANGES_REQUESTED** [review]

File: .github/workflows/continuous-integration.yml (hunk 2 above)
Comment (rafayet-monon) [reply .github/workflows/continuous-integration.yml]: The main reason I put it there because if `npm install` fails it will not run `ESLint` and `Danger` won't be able to read `ESLint` error if it's not run. But I guess if `npm install` fails nothing will run so it can be removed!

File: src/App.test.js (hunk 5 above)
Comment (rafayet-monon) [reply src/App.test.js]: I guess there is no method `context` in `jest` - https://jestjs.io/docs/en/api#methods

Comment (namtx) [reply src/App.test.js]: Ah, yes, thanks for referencing

Comment (rafayet-monon) [reply src/App.test.js]: Updated with the reviewed changes.

File: .github/workflows/continuous-integration.yml (hunk 2 above)
Comment (namtx) [reply .github/workflows/continuous-integration.yml]: should we remove it from other places?

File: dangerfile.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    let file_path = lint_obj.filePath;
+    let line_no = lint_obj.messages[0].line;
+    let danger_message = lint_obj.messages[0].message;
+
+    let full_message = `Path: ${file_path} - Line: ${line_no} - message: ${danger_message}`;
```
Comment (namtx) [inline dangerfile.js]:
```suggestion
    const full_message = `Path: ${file_path} - Line: ${line_no} - message: ${danger_message}`;
```
I think

File: .github/workflows/continuous-integration.yml (hunk 2 above)
Comment (rafayet-monon) [reply .github/workflows/continuous-integration.yml]: If we do that and then when somehow the `Run ESLint` fails, Github Actions will just skip all the next steps. The `Danger` process won't run either so it won't be able to make comments in the PR with the lint problems.  By providing `if: always()` in every other step makes sure that it runs even if the previous step fails. This way `Danger` can pick up all the problems and comment accordingly. https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions

File: src/App.js (hunk 4 above)
Comment (rafayet-monon) [reply src/App.js]: Updated!

File: .github/workflows/continuous-integration.yml (hunk 2 above)
Comment (namtx) [reply .github/workflows/continuous-integration.yml]: oh, get it. Thanks for explaining

**Review by namtx — APPROVED** [review]
LGTM!

File: README.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
+# SURVEYOR
```
Comment (Lahphim) [inline README.md:1]: @rafayet-monon We have `git-template` 🙌 
What do you think if we have the same format like them? 🤔
https://github.com/nimblehq/git-template/blob/master/README.md

What we need to add more is about the `License` and `About` section. 👍

File: package.json [hunk 8]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "surveyor",
+  "version": "0.1.0",
+  "private": true,
+  "dependencies": {
```
Comment (Lahphim) [inline package.json]: @rafayet-monon We can move `dependencies ` block to address before `devDependencies` for making it easier to see 😁

File: package.json [hunk 9]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "surveyor",
+  "version": "0.1.0",
+  "private": true,
```
Comment (Lahphim) [inline package.json:4]: @rafayet-monon is this mean private repository? 🤔

File: package.json [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  "dependencies": {
+    "node-sass": "^4.14.1",
+    "react": "^16.13.1",
```
Comment (Lahphim) [inline package.json]: @rafayet-monon We can set some packages to install the specific latest version here.
For example here `"react": "16.13.1"`.
Because another collaborator in the in this project will get the same version as your local machine 😁

File: dangerfile.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+const {fail} =  require("danger")
```
Comment (Lahphim) [inline dangerfile.js]: @rafayet-monon Should we make this with this format `const { fail } =  require('danger')`?

File: dangerfile.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+const fs = require("fs")
+const jest = require('danger-plugin-jest/dist/index.js')
+const stylelint = require('danger-plugin-stylelint')
+
+// ESLINT
+const esLintFile = "eslint-output.json";
```
Comment (Lahphim) [inline dangerfile.js]: @rafayet-monon Prefer single quotes. 🙌

Screen Shot 2563-08-19 at 11 21 03

File: public/manifest.json [hunk 13]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (Lahphim) [inline public/manifest.json:1]: @rafayet-monon Should we update this file to have just the only thing you need and its message about your project?
Because you're using this one which is not related with your project anymore.
```
"short_name": "React App",
"name": "Create React App Sample",
```

File: package.json (hunk 9 above)
Comment (rafayet-monon) [reply package.json:4]: From here - https://docs.npmjs.com/files/package.json#private

It is a way to prevent accidental publication of private repositories to any registry. i.e. `npm registry` or any internal registry. I guess it's not associated with Git repositories.

File: README.md (hunk 7 above)
Comment (rafayet-monon) [reply README.md:1]: Updated requested changes!

File: package.json [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    "react-scripts": "3.4.3"
+  },
+  "devDependencies": {
+    "@testing-library/jest-dom": "4.2.4",
+    "@testing-library/react": "9.5.0",
+    "@testing-library/user-event": "7.2.1",
+    "babel-eslint": "10.1.0",
+    "cross-env": "7.0.2",
+    "danger": "10.3.0",
+    "danger-plugin-jest": "^1.3.0",
```
Comment (Lahphim) [inline package.json]: @rafayet-monon At this block also 🙌
We can remove `^` out.

Comment (rafayet-monon) [reply package.json]: Fixed 🚀

**Review by Lahphim — APPROVED** [review]
